metis-indexing/src/main/java/eu/europeana/indexing/FullBeanPublisher.java

```diff
@@ -159,4 +189,68 @@ private void publishToSolr(RdfWrapper rdf, FullBeanImpl fullBean) throws Indexin
       throw new RecordRelatedIndexingException(SOLR_SERVER_PUBLISH_ERROR, e);
     }
   }
+
+  private String searchMatchingRecordForRedirection(String recordIdIncludingDatasetId,
+      List<String> datasetIdsForRedirection)
+      throws IndexerRelatedIndexingException, RecordRelatedIndexingException {
+    //Check if record with same record identifier exists
+    final Map<String, String> queryParamMap = new HashMap<>();
+    queryParamMap.put("q",
+        String.format("%s:%s", EdmLabel.EUROPEANA_ID.toString(), recordIdIncludingDatasetId));
+    queryParamMap.put("fl", EdmLabel.EUROPEANA_ID.toString());
+
+    SolrDocumentList documents = getSolrDocuments(queryParamMap);
+    if (!CollectionUtils.isEmpty(documents)) {
+      return null;
+    }
+
+    //Check matches with older dataset identifiers
+    if (!CollectionUtils.isEmpty(datasetIdsForRedirection)) {
+      //The incoming structure of the identifier is /datasetId/recordId
+      final String[] splitRecordIdentifier = recordIdIncludingDatasetId.split("/");
+      String datasetId = splitRecordIdentifier[1];
+      String recordId = splitRecordIdentifier[2];
+
+      final String combinedQueryForRedirectedDatasetIds = datasetIdsForRedirection.stream()
+          .map(datasetIdForRedirection -> String
+              .format("%s:/%s/%s", EdmLabel.EUROPEANA_ID.toString(), datasetIdForRedirection,
+                  recordId)).collect(Collectors.joining(" OR "));
+      queryParamMap.put("q", combinedQueryForRedirectedDatasetIds);
+    }
+
+    documents = getSolrDocuments(queryParamMap);
+    if (!CollectionUtils.isEmpty(documents)) {
+      //Get identifier of the first document found
+      return (String) documents.get(0).getFieldValue(EdmLabel.EUROPEANA_ID.toString());
+    }
+    return null;
+  }
+
+  private SolrDocumentList getSolrDocuments(Map<String, String> queryParamMap)
+      throws IndexerRelatedIndexingException, RecordRelatedIndexingException {
+    MapSolrParams queryParams = new MapSolrParams(queryParamMap);
+    QueryResponse response;
+    try {
+      response = solrServer.query(queryParams);
+    } catch (SolrServerException e) {
+      throw new IndexerRelatedIndexingException(SOLR_SERVER_SEARCH_ERROR, e);
+    } catch (IOException e) {
+      throw new RecordRelatedIndexingException(SOLR_SERVER_SEARCH_ERROR, e);
+    }
+    return response.getResults();
+  }
+
+  private void createRedirection(String newIdentifier, String oldIdentifier,
+      Date recordRedirectDate) {
+    final RecordRedirect recordRedirect = new RecordRedirect(newIdentifier, oldIdentifier,
+        recordRedirectDate);
+    recordRedirectDao.createUpdate(recordRedirect);
+    //Update the previous redirect item in db that has newId == oldIdentifier
+    final RecordRedirect recordRedirectByNewId = recordRedirectDao
+        .getRecordRedirectByNewId(oldIdentifier);
```

Comment:
There could be multiple of these, I think. We should update all of them.

Reply:
No that shouldn't be possible